File: general_analysis/plotIllumAndObjectColors.R
colormatchPlanck <- function(spect, wlns) {
	cmfData <- read.csv('C:\\Users\\me\\Documents\\source_code\\MATLAB\\matlabHyper\\ciexyz31_1.csv')

	wavelength.cmf <- cmfData[ , 1]

	x.bar <- cmfData[ , 2]
	y.bar <- cmfData[ , 3]
	z.bar <- cmfData[ , 4]

	x.cmf.fun <- splinefun(x = wavelength.cmf, y = x.bar)
	x.cmf <- x.cmf.fun(wlns)
	x.cmf[x.cmf < 0.0] <- 0.0
	y.cmf.fun <- splinefun(x = wavelength.cmf, y = y.bar)
	y.cmf <- y.cmf.fun(wlns)
	y.cmf[y.cmf < 0.0] <- 0.0
	z.cmf.fun <- splinefun(x = wavelength.cmf, y = z.bar)
	z.cmf <- z.cmf.fun(wlns)
	z.cmf[z.cmf < 0.0] <- 0.0

	XYZ <- c(0, 0, 0)

	stepSize <- c(diff(wlns), 0)
	x.corrCMF <- 683 * x.cmf * stepSize
	y.corrCMF <- 683 * y.cmf * stepSize
	z.corrCMF <- 683 * z.cmf * stepSize

	XYZ[1] <- spect %*% x.corrCMF
	XYZ[2] <- spect %*% y.corrCMF
	XYZ[3] <- spect %*% z.corrCMF

	# monxyY <- read.csv('OLEDxyY.csv')
	# RGB <- XYZ2RGB(XYZ, monxyY)

	return(XYZ)
}

XYZtoxyY <- function(XYZ) {
	s <- sum(XYZ)
	xyY <- c(XYZ[1]/s, XYZ[2]/s, XYZ[2])
	return(xyY)
}

# load distributions for illum and object colors
lightredrefl <- read.table('../base_stimuli/spectra/munsell_red_lightest.spd', sep=' ')
lightbluerefl <- read.table('../base_stimuli/spectra/munsell_blue_lightest.spd', sep=' ')
lightgreenrefl <- read.table('../base_stimuli/spectra/munsell_green_lightest.spd', sep=' ')
lightyellowrefl <- read.table('../base_stimuli/spectra/munsell_yellow_lightest.spd', sep=' ')

wlns <- lightredrefl[ , 1]

lightredxyy <- XYZtoxyY(colormatchPlanck(lightredrefl[ , 2], wlns))
lightgreenxyy <- XYZtoxyY(colormatchPlanck(lightgreenrefl[ , 2], wlns))
lightbluexyy <- XYZtoxyY(colormatchPlanck(lightbluerefl[ , 2], wlns))
lightyellowxyy <- XYZtoxyY(colormatchPlanck(lightyellowrefl[ , 2], wlns))

bluespectra <- read.table('../base_stimuli/spectra/de_brighter_illum_blue.spd', sep=' ')
yellowspectra <- read.table('../base_stimuli/spectra/de_brighter_illum_yellow.spd', sep=' ')

wlns <- bluespectra[ , 1]

bluesxyy <- XYZtoxyY(colormatchPlanck(bluespectra[ , 2], wlns))
yellowsxyy <- XYZtoxyY(colormatchPlanck(yellowspectra[ , 2], wlns))

# load distributions for filter matching

filterredrefl <- read.table('../base_stimuli/spectra/munsell_red_better.spd', sep=' ')
filterbluerefl <- read.table('../base_stimuli/spectra/munsell_blue_better.spd', sep=' ')
filtergreenrefl <- read.table('../base_stimuli/spectra/munsell_green_better.spd', sep=' ')
filteryellowrefl <- read.table('../base_stimuli/spectra/munsell_yellow_better.spd', sep=' ')

wlns <- lightredrefl[ , 1]

filterredxyy <- XYZtoxyY(colormatchPlanck(filterredrefl[ , 2], wlns))
filtergreenxyy <- XYZtoxyY(colormatchPlanck(filtergreenrefl[ , 2], wlns))
filterbluexyy <- XYZtoxyY(colormatchPlanck(filterbluerefl[ , 2], wlns))
filteryellowxyy <- XYZtoxyY(colormatchPlanck(filteryellowrefl[ , 2], wlns))

# grab pre-computed xyY coords of CIE lasso and planckian locus for plotting
lassocs <- read.csv('C:\\Users\\me\\Documents\\source_code\\MATLAB\\xyYcoords.csv')
planck <- read.csv('C:\\Users\\me\\Documents\\source_code\\MATLAB\\planck.csv')

mypalette()

pdf("../figures/refl_illum_coords.pdf", width = 9, height = 4)
mypar(mfrow=c(1, 2), pty="s", cex.axis=1, cex = 1)
myplot(NA, NA, xlim=c(0, 0.8), ylim=c(0, 0.9))

mylines(lassocs[ , 1], lassocs[ , 2])
mylines(planck[ , 1], planck[ , 2])

mypoints(lightredxyy[1], lightredxyy[2], col=4, pch=16, cex=1.1)
mypoints(lightgreenxyy[1], lightgreenxyy[2], col=5, pch=16, cex=1.1)
mypoints(lightbluexyy[1], lightbluexyy[2], col=6, pch=16, cex=1.1)
mypoints(lightyellowxyy[1], lightyellowxyy[2], col=7, pch=16, cex=1.1)

mypoints(filterredxyy[1], filterredxyy[2], col=4, pch=17, cex=1.1)
mypoints(filtergreenxyy[1], filtergreenxyy[2], col=5, pch=17, cex=1.1)
mypoints(filterbluexyy[1], filterbluexyy[2], col=6, pch=17, cex=1.1)
mypoints(filteryellowxyy[1], filteryellowxyy[2], col=7, pch=17, cex=1.1)

mypoints(bluesxyy[1], bluesxyy[2], col=6, pch=8, cex=0.8)
mypoints(yellowsxyy[1], yellowsxyy[2], col=1, pch=8, cex=0.8)

myaxis(1, at = seq(0, 0.8, 0.8/5))
myaxis(2, at = round(seq(0, 0.89, 0.89/5), digits=2))
mytitle(xlab = "x", ylab = "y")

myplot(NA, NA, xlim=c(380, 830), ylim=c(0, max(bluespectra[, 2])+1))
mylines(bluespectra[, 1], bluespectra[, 2], col=6)
mylines(yellowspectra[, 1], yellowspectra[, 2], col=1)

myaxis(1, at = round(seq(380, 830, (830 - 380)/5)))
myaxis(2, at = round(seq(min(bluespectra[, 2]), max(bluespectra[, 2]), (max(bluespectra[, 2]) - min(bluespectra[, 2]))/5)))
mytitle(xlab = "Wavelength (nm)", ylab = expression(paste("Radiance ", ~~ (frac(W, str %*% m^{2})))))
dev.off()

File: general_analysis/plotIllumAndObjectColors.m
clear all;
close all;
clc;

% Change default axes fonts.
set(0,'DefaultAxesFontSize', 15);

% Change default text fonts.
set(0,'DefaultTextFontSize', 15);

object_colors = zeros(8, 3);

% load distributions for illum and object colors
lightredrefl = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/munsell_red_lightest.spd');
lightbluerefl = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/munsell_blue_lightest.spd');
lightgreenrefl = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/munsell_green_lightest.spd');
lightyellowrefl = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/munsell_yellow_lightest.spd');

[lightredrgb,lightredx] = colormatchPlanck(lightredrefl(:,2), lightredrefl(:,1));
lightredrgb = lightredrgb./650;
object_colors(1, :) = lightredrgb
lightredxyy = [lightredx(1)/sum(lightredx) lightredx(2)/sum(lightredx) lightredx(2)];

[lightgreenrgb,lightgreenx] = colormatchPlanck(lightgreenrefl(:,2), lightgreenrefl(:,1));
lightgreenrgb = lightgreenrgb./650;
object_colors(2, :) = lightgreenrgb;
lightgreenxyy = [lightgreenx(1)/sum(lightgreenx) lightgreenx(2)/sum(lightgreenx) lightgreenx(2)];

[lightbluergb,lightbluex] = colormatchPlanck(lightbluerefl(:,2), lightbluerefl(:,1));
lightbluergb = lightbluergb./650;
object_colors(3, :) = lightbluergb;
lightbluexyy = [lightbluex(1)/sum(lightbluex) lightbluex(2)/sum(lightbluex) lightbluex(2)];

[lightyellowrgb,lightyellowx] = colormatchPlanck(lightyellowrefl(:,2), lightyellowrefl(:,1));
lightyellowrgb = lightyellowrgb./650;
object_colors(4, :) = lightyellowrgb;
lightyellowxyy = [lightyellowx(1)/sum(lightyellowx) lightyellowx(2)/sum(lightyellowx) lightyellowx(2)];

% plot light transmissions - one at a time

figure(3);
clf;
hold on;
xlim([lightredrefl(1,1), lightredrefl(end,1)]);
ylim([0,100]);
ylabel('Transmission (%)');
xlabel('Wavelength (nm)');
plot(lightredrefl(:,1), 100.0*lightredrefl(:,2), 'r-', 'LineWidth', 2, 'Color', lightredrgb);
axis square;
export_fig('light_red_tranmission.pdf', '-dpdf');

figure(4);
clf;
hold on;
xlim([lightredrefl(1,1), lightredrefl(end,1)]);
ylim([0,100]);
ylabel('Transmission (%)');
xlabel('Wavelength (nm)');
plot(lightgreenrefl(:,1), 100.0*lightgreenrefl(:,2), 'g-', 'LineWidth', 2, 'Color', lightgreenrgb);
axis square;
export_fig('light_green_tranmission.pdf', '-dpdf');

figure(5);
clf;
hold on;
xlim([lightredrefl(1,1), lightredrefl(end,1)]);
ylim([0,100]);
ylabel('Transmission (%)');
xlabel('Wavelength (nm)');
plot(lightbluerefl(:,1), 100.0*lightbluerefl(:,2), 'b-', 'LineWidth', 2, 'Color', lightbluergb);
axis square;
export_fig('light_blue_tranmission.pdf', '-dpdf');

figure(6);
clf;
hold on;
xlim([lightredrefl(1,1), lightredrefl(end,1)]);
ylim([0,100]);
ylabel('Transmission (%)');
xlabel('Wavelength (nm)');
plot(lightyellowrefl(:,1), 100.0*lightyellowrefl(:,2), 'y-', 'LineWidth', 2, 'Color', lightyellowrgb);
axis square;
export_fig('light_yellow_tranmission.pdf', '-dpdf');

darkredrefl = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/munsell_red_lighter.spd');
darkbluerefl = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/munsell_blue_lighter.spd');
darkgreenrefl = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/munsell_green_lighter.spd');
darkyellowrefl = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/munsell_yellow_lighter.spd');

[darkredrgb,darkredx] = colormatchPlanck(darkredrefl(:,2), darkredrefl(:,1));
darkredrgb = darkredrgb./650;
object_colors(5, :) = darkredrgb;
darkredxyy = [darkredx(1)/sum(darkredx) darkredx(2)/sum(darkredx) darkredx(2)];

[darkgreenrgb,darkgreenx] = colormatchPlanck(darkgreenrefl(:,2), darkgreenrefl(:,1));
darkgreenrgb = darkgreenrgb./650;
object_colors(6, :) = darkgreenrgb;
darkgreenxyy = [darkgreenx(1)/sum(darkgreenx) darkgreenx(2)/sum(darkgreenx) darkgreenx(2)];

[darkbluergb,darkbluex] = colormatchPlanck(darkbluerefl(:,2), darkbluerefl(:,1));
darkbluergb = darkbluergb./650;
object_colors(7, :) = darkbluergb;
darkbluexyy = [darkbluex(1)/sum(darkbluex) darkbluex(2)/sum(darkbluex) darkbluex(2)];

[darkyellowrgb,darkyellowx] = colormatchPlanck(darkyellowrefl(:,2), darkyellowrefl(:,1));
darkyellowrgb = darkyellowrgb./650;
object_colors(8, :) = darkyellowrgb;
darkyellowxyy = [darkyellowx(1)/sum(darkyellowx) darkyellowx(2)/sum(darkyellowx) darkyellowx(2)];

% plot dark transmissions - one at a time

figure(3);
clf;
hold on;
axis square;
xlim([lightredrefl(1,1), lightredrefl(end,1)]);
ylim([0,100]);
ylabel('Transmission (%)');
xlabel('Wavelength (nm)');
plot(darkredrefl(:,1), 100.0*darkredrefl(:,2), 'r-', 'LineWidth', 2, 'Color', darkredrgb);
axis square;
export_fig('dark_red_tranmission.pdf', '-dpdf');

figure(4);
clf;
hold on;
axis square;
xlim([lightredrefl(1,1), lightredrefl(end,1)]);
ylim([0,100]);
ylabel('Transmission (%)');
xlabel('Wavelength (nm)');
plot(darkgreenrefl(:,1), 100.0*darkgreenrefl(:,2), 'g-', 'LineWidth', 2, 'Color', darkgreenrgb);
axis square;
export_fig('dark_green_tranmission.pdf', '-dpdf');

figure(5);
clf;
hold on;
axis square;
xlim([lightredrefl(1,1), lightredrefl(end,1)]);
ylim([0,100]);
ylabel('Transmission (%)');
xlabel('Wavelength (nm)');
plot(darkbluerefl(:,1), 100.0*darkbluerefl(:,2), 'b-', 'LineWidth', 2, 'Color', darkbluergb);
axis square;
export_fig('dark_blue_tranmission.pdf', '-dpdf');

figure(6);
clf;
hold on;
axis square;
xlim([lightredrefl(1,1), lightredrefl(end,1)]);
ylim([0,100]);
ylabel('Transmission (%)');
xlabel('Wavelength (nm)');
plot(darkyellowrefl(:,1), 100.0*darkyellowrefl(:,2), 'y-', 'LineWidth', 2, 'Color', darkyellowrgb);
axis square;
export_fig('dark_yellow_tranmission.pdf', '-dpdf');

redspectra = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/de_brighter_illum_red.spd');
bluespectra = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/de_brighter_illum_blue.spd');
greenspectra = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/de_brighter_illum_green.spd');
yellowspectra = dlmread('/Users/rje/my_docs/simpler_pathmarcher/spectra/de_brighter_illum_yellow.spd');

illum_colors = zeros(2, 3);

[redsrgb,redsx] = colormatchPlanck(redspectra(:,2), redspectra(:,1));
redsrgb = redsrgb./2500;
redsxyy = [redsx(1)/sum(redsx) redsx(2)/sum(redsx) redsx(2)];
[greensrgb,greensx] = colormatchPlanck(greenspectra(:,2), greenspectra(:,1));
greensrgb = greensrgb./2500;
greensxyy = [greensx(1)/sum(greensx) greensx(2)/sum(greensx) greensx(2)];

[bluesrgb,bluesx] = colormatchPlanck(bluespectra(:,2), bluespectra(:,1));
bluesrgb = bluesrgb./2500;
illum_colors(1, :) = bluesrgb;
bluesxyy = [bluesx(1)/sum(bluesx) bluesx(2)/sum(bluesx) bluesx(2)];

[yellowsrgb,yellowsx] = colormatchPlanck(yellowspectra(:,2), yellowspectra(:,1));
yellowsrgb = yellowsrgb./2500;
illum_colors(2, :) = yellowsrgb;
yellowsxyy = [yellowsx(1)/sum(yellowsx) yellowsx(2)/sum(yellowsx) yellowsx(2)];

save('object_illum_colors_for_plot.mat', 'object_colors', 'illum_colors');

pause;

% plot illum spectra - one at a time

% note that somehow, someway, the green and red spectra were accidentally switched into
% different files. as usual, i have no idea how...

% green spectrum
figure(3);
clf;
hold on;
axis square;
ylim([0,14]);
xlabel('Wavelength (nm)');
ylabel('Radiance (W/(sr * m^{2}))');
plot(redspectra(:,1), redspectra(:,2), 'r-', 'LineWidth', 2, 'Color', redsrgb);
export_fig('illum_green_spectra.pdf', '-dpdf');

% red spectrum
figure(4);
clf;
hold on;
axis square;
ylim([0,14]);
xlabel('Wavelength (nm)');
ylabel('Radiance (W/(sr * m^{2}))');
plot(greenspectra(:,1), greenspectra(:,2), 'g-', 'LineWidth', 2, 'Color', greensrgb);
export_fig('illum_red_spectra.pdf', '-dpdf');

figure(5);
clf;
hold on;
axis square;
ylim([0,14]);
xlabel('Wavelength (nm)');
ylabel('Radiance (W/(sr * m^{2}))');
plot(bluespectra(:,1), bluespectra(:,2), 'b-', 'LineWidth', 2, 'Color', bluesrgb);
export_fig('illum_blue_spectra.pdf', '-dpdf');

figure(6);
clf;
hold on;
axis square;
ylim([0,14]);
xlabel('Wavelength (nm)');
ylabel('Radiance (W/(sr * m^{2}))');
plot(yellowspectra(:,1), yellowspectra(:,2), 'y-', 'LineWidth', 2, 'Color', yellowsrgb);
export_fig('illum_yellow_spectra.pdf', '-dpdf');

% grab pre-computed xyY coords of CIE lasso and planckian locus for plotting
lassocs = csvread('xyYcoords.csv');
planck = csvread('planck.csv');

% first plot showing transmission colors (only need to plot lighter or darker, since they have
% same chroma, but different luminance)
figure(1);
clf;
hold on;
plot(lassocs(:,1), lassocs(:,2),'k-');
plot(planck(:,1), planck(:,2),'k-');
plot(lightredxyy(1), lightredxyy(2), 'ro', 'MarkerFaceColor', lightredrgb, 'MarkerEdgeColor', lightredrgb);
plot(lightgreenxyy(1), lightgreenxyy(2), 'go', 'MarkerFaceColor', lightgreenrgb, 'MarkerEdgeColor', lightgreenrgb);
plot(lightbluexyy(1), lightbluexyy(2), 'bo', 'MarkerFaceColor', lightbluergb, 'MarkerEdgeColor', lightbluergb);
plot(lightyellowxyy(1), lightyellowxyy(2), 'yo', 'MarkerFaceColor', lightyellowrgb, 'MarkerEdgeColor', lightyellowrgb);
axis([0 1 0 1]);
axis square;
xlabel('x');
ylabel('y');
title('"Chromaticity coordinates" of transmission distributions');
axis square

export_fig('transmission_colors_xyy.pdf', '-dpdf');

% second plot showing blue-yellow illuminant colors
figure(2);
clf;
hold on;
plot(lassocs(:,1), lassocs(:,2),'k-');
plot(planck(:,1), planck(:,2),'k-');
plot(bluesxyy(1), bluesxyy(2), 'bo', 'MarkerFaceColor', bluesrgb, 'MarkerEdgeColor', bluesrgb);
plot(yellowsxyy(1), yellowsxyy(2), 'yo', 'MarkerFaceColor', yellowsrgb, 'MarkerEdgeColor', yellowsrgb);
axis([0 1 0 1]);
axis square;
xlabel('x');
ylabel('y');
title('Chromaticity coordinates of "blue-yellow" illuminants');
axis square

export_fig('blue_yellow_illum_xyy.pdf', '-dpdf');

% third plot showing red-green illuminant colors
figure(2);
clf;
hold on;
plot(lassocs(:,1), lassocs(:,2),'k-');
plot(planck(:,1), planck(:,2),'k-');
% red and green were accidentally swapped
plot(redsxyy(1), redsxyy(2), 'ro', 'MarkerFaceColor', redsrgb, 'MarkerEdgeColor', redsrgb);
plot(greensxyy(1), greensxyy(2), 'go', 'MarkerFaceColor', greensrgb, 'MarkerEdgeColor', greensrgb);
axis([0 1 0 1]);
axis square;
xlabel('x');
ylabel('y');
title('Chromaticity coordinates of "red-green" illuminants');
axis square

export_fig('red_green_illum_xyy.pdf', '-dpdf');
